File: Editorial/2464_Minimum_Subarrays_in_a_Valid_Split/Go/1_Approach_Dynamic_Programming.go
const INF = 1e9

func validSubarraySplit(nums []int) int {
	n := len(nums)
	dp := make([]int, n+1)
	for i := 0; i <= n; i++ {
		dp[i] = INF
	}
	dp[0] = 0
	for i := 1; i <= n; i++ {
		for j := 1; j <= i; j++ {
			if gcd(nums[i-1], nums[j-1]) > 1 {
				dp[i] = min(dp[i], dp[j-1]+1)
			}
		}
	}
	if dp[n] == INF {
		return -1
	}
	return dp[n]
}

func gcd(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

File: Editorial/2464_Minimum_Subarrays_in_a_Valid_Split/CPP/1_Approach_Dynamic_Programming.cpp
class Solution {
public:
    static constexpr int INF = 1e9;
    int validSubarraySplit(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n + 1, INF);
        dp[0] = 0;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= i; j++) {
                if (__gcd(nums[i - 1], nums[j - 1]) > 1) {
                    dp[i] = min(dp[i], dp[j - 1] + 1);
                }
            }
        }
        return dp[n] == INF ? -1 : dp[n];
    }
};
